### ISSUE TITLE
Support inline definitions of suppressions because relative paths dont work well with 3rd party tools

### ISSUE THREAD

**Issue by msteiger** [checkstyle/checkstyle#1051]
Currently, suppressions need to be defined in a separate xml file. I think it would be useful if that was not necessary.

An example: Since Gradle 2.2 [config files can be specified as artifacts](https://gradle.org/docs/current/dsl/org.gradle.api.plugins.quality.CheckstyleExtension.html). Thus, they can be downloaded from a Maven repository just like any jar file and are stored somewhere on disk. The problem with the suppressions file is that the `checkstyle.xml` needs to find it relative to it's own location, [which is tricky](https://discuss.gradle.org/t/checkstyle-suppressions-not-working-with-config-textresource-fromarchiveentry/9198).

The [definition of suppressions is simple](http://www.puppycrawl.com/dtds/suppressions_1_1.dtd) - supporting that in the main file should be easy and keep the complexity/length of the file almost as low as it is now.
##

**Comment by romani** [comment]
I agree in general, but I do not see good solution for this now (but I did not thought a lot about this).

**Comment by romani** [comment]
from SForge feature request 520:
```
 I've been able to distribute checks to my students using an absolute pathname for the location of the standard suppressions file, but if it's a Windows pathname, my Mac chokes on it, and vice versa.

If you don't want to mess up the DTD, it could look like:
<module name="SuppressionFilter">
<property name="literal" value="<suppress files=\"Test.*.java\"
checks=\"MagicNumber\"/><suppress other-rules-go-here/>"/>
</module>

This is a little more error-prone and harder to use than the more user-obvious
    <module name="SuppressionFilter">
        <suppress files="Test.*.java" checks="MagicNumber"/>
        <suppress other-rules-go-here/>
    </module>

In fact, it would be neat if SuppressionFilter allowed BOTH a "file" property and an arbitrary number of <suppress> tags; that way a whole site could have a baseline suppression filter file, and individual users or projects could add other suppression filters to it.  Even better would be to allow multiple files as well, of course...
```

**Comment by ChristianWulf** [comment]
PMD solves this issue very conveniently:
```xml
<?xml version="1.0"?>
<ruleset name="Kieker PMD ruleset" xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 http://pmd.sourceforge.net/ruleset_2_0_0.xsd"
	xsi:noNamespaceSchemaLocation="http://pmd.sourceforge.net/ruleset_2_0_0.xsd">
	<description>
    This ruleset is checked for Kieker automatically.
  </description>

	<!-- exclude generated code -->
	<exclude-pattern>.*/src-gen/.*</exclude-pattern>
	<exclude-pattern>.*/test-gen/.*</exclude-pattern>
	<exclude-pattern>.*/examples/userguide/.*</exclude-pattern>

  ...
```

**Comment by ChristianWulf** [comment]
I have found a practical solution which already works with all Checkstyle versions:
```xml
	<module name="BeforeExecutionExclusionFileFilter">
		<property name="fileNamePattern" value=".*[/\\]src-gen[/\\].*"/>
	</module>

	<module name="BeforeExecutionExclusionFileFilter">
		<property name="fileNamePattern" value=".*test-gen.*" />
	</module>
```
For me, this is sufficient. It avoids passing the absolute file path of the suppressions.xml and thus works in all environments, with all build tools, and with all IDE Plugins.

**Comment by rnveach** [comment]
>  It avoids passing the absolute file path of the suppressions.xml 

Suppressions.xml doesn't require absolute path. It supports regular expression too. See our configuration as an example: https://github.com/checkstyle/checkstyle/blob/master/config/suppressions.xml#L13

> I have found a practical solution which already works with all Checkstyle

This is different than what the issue creator wants, so it is misleading saying this is a solution to the issue.
You are suppressing entire files, where creator just wants to define suppressions (probably by line or check) in the configuration file instead of an external file.

**Comment by ChristianWulf** [comment]
Me, and I think the author, too, complain about the path to the suppression. xml. This path must be absolute to work with Gradle, Eclipse Plugins etc. I do not talk about the contents of the suppression.xml. 

Checkstyle, in my opinion, does not resolve relative paths to external xml files very well. Everythink would be easy and convinient if relative paths would be resolved relative to the folder of the config file. Since this is not implemented, the only current way of declaring suppressions in a tool-independent way is to avoid such file path indications.

**Comment by o-nix** [comment]
Oh my it is 2+ years unresolved proposal...
Having problems with properties + suppression file path, just wanted to move everything under `checkstyle.xml`, thought it would be easy...

**Comment by romani** [comment]
@ChristianWulf , sorry but I removed your two comments as they are off topic in such complicated issue.

I updated by first comment to explain some misunderstanding of modules in checkstyle. Please re-read.
All plugins have almost all possibilities to configure and run checkstyle in proper way. But yes, this is not simple issue for them sometime. 

>  This path must be absolute to work with Gradle, Eclipse Plugins etc.

it may be path from classpath, 
from http://checkstyle.sourceforge.net/config_filters.html#SuppressionFilter
>    as a filesystem location
    if no file found, and the location starts with either http:// or https://, then it is interpreted as a URL
    if no file found, then passed to the ClassLoader.getResource() method.

Please consider to use configs from resources, bundle all checkstyle configs to jar and it should work, if your build tool system allow this.
One more alternative is URLs, ... small performance hit to download file, dependency to network ... but could be considered as workaround.

**Comment by johanhammar** [comment]
I have a different use case which might, or might not be related to this. We provision a "global" checkstyle configuration file which is located outside of any project directory. This helps us to enforce a good coding style in all our projects. However it would be nice if we could put a suppressions file in any project i.e. in a config directory or in the root of the project. Unfortunately this seems to be quite tricky to get working, especially in a multi module project (Maven).

**Comment by rnveach** [comment]
@johanhammar 
> it would be nice if we could put a suppressions file in any project i.e. in a config directory or in the root of the project

Check out how we set suppression paths by variables at https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_checks.xml#L34-L35 . This is what you should be using as each project needs to specify these variables.

**Comment by johanhammar** [comment]
@rnveach It seems near impossible to define a variable that points to the root of the project and getting eclipse-cs or the maven-checkstyle-plugin to find the project specific suppressions file. 

This might be a bit off topic. Should we move to the mailing list?

**Comment by rnveach** [comment]
@johanhammar Feel free to open a mailing list if you need help.

**Comment by rnveach** [comment]
I am closing this issue. We have http://checkstyle.sourceforge.net/config_filters.html#SuppressionXpathSingleFilter and we have a new issue for SuppressionSingleFilter at https://github.com/checkstyle/checkstyle/issues/6402 .

Feel free to reopen the issue these filters are not enough for your needs and why.
